### PR TITLE
feat(*): allow to select auth providers + add multiworkspace with subdomain management

### PR DESCRIPTION
**Pull request by AMoreaux**
## Summary
Add support for multi-workspace feature and adjust configurations and states accordingly.
- Introduced new state isMultiWorkspaceEnabledState.
- Updated ClientConfigProviderEffect component to handle multi-workspace.
- Modified GraphQL schema and queries to include multi-workspace related configurations.
- Adjusted server environment variables and their respective documentation to support multi-workspace toggle.
- Updated server-side logic to handle new multi-workspace configurations and conditions.

### CODE REVIEW

**Review by martmull — COMMENTED** [review]
LGTM, one comment about default value

File: packages/twenty-server/src/engine/core-modules/environment/environment-variables.ts [hunk 1]
```diff
@@ -232,6 +232,11 @@ export class EnvironmentVariables {
   @IsOptional()
   ENTERPRISE_KEY: string;
 
+  @CastToBoolean()
+  @IsOptional()
+  @IsBoolean()
+  IS_MULTIWORKSPACE_ENABLED = false;
```
[the remark this answers is not in the document]
Comment (AMoreaux) [reply packages/twenty-server/src/engine/core-modules/environment/environment-variables.ts:238]: Done 👌

**Comment by AMoreaux** [comment]
⚠️ Before merging it wait for the PR that adds subdomains.

**Comment by AMoreaux** [comment]
@martmull I converted this PR to a draft to avoid the merge since we must wait for the review on child PR to have the full feature.

File: packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts [hunk 2]
```diff
@@ -297,11 +297,16 @@ export class SignInUpService {
     lastName: string;
     picture: SignInUpServiceInput['picture'];
   }) {
-    if (this.environmentService.get('IS_SIGN_UP_DISABLED')) {
-      throw new AuthException(
-        'Sign up is disabled',
-        AuthExceptionCode.FORBIDDEN_EXCEPTION,
-      );
+    if (!this.environmentService.get('IS_MULTIWORKSPACE_ENABLED')) {
+      const numberOfWorkspaces = await this.workspaceRepository.count();
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts:301]: workspacesCount

**Review by charlesBochet — APPROVED** [review]

File: packages/twenty-front/src/modules/auth/sign-in-up/components/SignInUpWorkspaceScopeForm.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+export const SignInUpWorkspaceScopeForm = () => {
+  const [authProviders] = useRecoilState(authProvidersState);
+
+  const { form } = useSignInUpForm();
+  const { handleResetPassword } = useHandleResetPassword();
+
+  const { signInUpStep, continueWithEmail } = useSignInUp(form);
+
+  const checkAuthProviders = useCallback(() => {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/auth/sign-in-up/components/SignInUpWorkspaceScopeForm.tsx:28]: move to SignInUpWorkspaceScopeFormEffect

File: packages/twenty-front/src/modules/workspace/components/WorkspaceProviderEffect.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    },
+    onError: (err) => {
+      // eslint-disable-next-line no-console
+      console.error(err);
+    },
+  });
+
+  useEffect(() => {
+    try {
+      if (isDefined(workspacePublicData?.logo)) {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/workspace/components/WorkspaceProviderEffect.tsx:31]: see if can leverage react-helmet

File: packages/twenty-front/src/modules/auth/types/signInUpMode.type.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
+export enum SignInUpMode {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/auth/types/signInUpMode.type.ts:1]: filenaming

File: packages/twenty-front/src/modules/settings/security/components/SettingsSSOIdentitiesProvidersListCard.tsx [hunk 6]
```diff
@@ -8,11 +8,14 @@ import { SettingsPath } from '@/types/SettingsPath';
...
 import { SettingsSSOIdentitiesProvidersListCardWrapper } from '@/settings/security/components/SettingsSSOIdentitiesProvidersListCardWrapper';
-import { SSOIdentitiesProvidersState } from '@/settings/security/states/SSOIdentitiesProviders.state';
 import isPropValid from '@emotion/is-prop-valid';
 import styled from '@emotion/styled';
-import { useRecoilValue } from 'recoil';
+import { useRecoilValue, useRecoilState } from 'recoil';
 import { IconKey } from 'twenty-ui';
+import { useListSsoIdentityProvidersByWorkspaceIdQuery } from '~/generated/graphql';
+import { SnackBarVariant } from '@/ui/feedback/snack-bar-manager/components/SnackBar';
+import { SSOIdentitiesProvidersState } from '@/settings/security/states/SSOIdentitiesProviders.state';
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/settings/security/components/SettingsSSOIdentitiesProvidersListCard.tsx:17]: no .state

File: packages/twenty-front/src/modules/settings/security/components/SettingsSecurityOptionsList.tsx [hunk 7]
```diff
@@ -2,24 +2,98 @@ import { currentWorkspaceState } from '@/auth/states/currentWorkspaceState';
...
   );
-  if (!isDefined(currentWorkspace)) {
-    throw new Error(
-      'The current workspace must be defined to edit its security options.',
-    );
-  }
 
   const [updateWorkspace] = useUpdateWorkspaceMutation();
 
+  const isValidAuthProvider = (
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/settings/security/components/SettingsSecurityOptionsList.tsx:35]: extract this to util so we can test it easily and we the typing complexity is "hidden"

File: packages/twenty-front/src/modules/settings/security/types/AuthProviders.type.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+export type AuthProvidersKeys =
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/settings/security/types/AuthProviders.type.ts:1]: GOOGLE
MICROSOFT

isGoogleAuthEnabled
isMicrosoftAuthEnabled

map(authEnablekyey: AuthProvidersKeys): GOOGLE | MICROSOFT (

)

Comment (charlesBochet) [reply packages/twenty-front/src/modules/settings/security/types/AuthProviders.type.ts:1]: not mandatory but I would have "avoided the complexity like this"

File: packages/twenty-front/src/utils/workspace-url.helper.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+  if (isDefined(searchParams)) {
+    Object.entries(searchParams).forEach(([key, value]) =>
+      url.searchParams.set(key, value),
+    );
+  }
+  return url.toString();
+};
+
+export const redirectToHome = () => {
```
Comment (charlesBochet) [inline packages/twenty-front/src/utils/workspace-url.helper.ts:51]: file naming + split into multi files

File: packages/twenty-server/src/engine/core-modules/auth/services/auth.service.ts [hunk 10]
```diff
@@ -98,15 +109,17 @@ export class AuthService {
...
     password?: string;
     firstName?: string | null;
     lastName?: string | null;
-    workspaceInviteHash?: string | null;
-    workspacePersonalInviteToken?: string | null;
+    workspaceInviteHash?: string;
+    workspacePersonalInviteToken?: string;
     picture?: string | null;
     fromSSO: boolean;
+    isAuthEnabled?: ReturnType<(typeof workspaceValidator)['isAuthEnabled']>;
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/auth/services/auth.service.ts:122]: isAuthProviderEnabled

File: packages/twenty-server/src/engine/core-modules/auth/controllers/google-auth.controller.ts [hunk 11]
```diff
@@ -53,6 +58,13 @@ export class GoogleAuthController {
       workspaceInviteHash,
       workspacePersonalInviteToken,
       fromSSO: true,
+      isAuthEnabled: workspaceValidator.isAuthEnabled(
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/auth/controllers/google-auth.controller.ts:61]: authProvider: 'google' | 'microsoft'

File: packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts [hunk 12]
```diff
@@ -137,47 +158,46 @@ export class SignInUpService {
...
+      new AuthException(
         'Workspace is not ready to welcome new members',
         AuthExceptionCode.FORBIDDEN_EXCEPTION,
-      );
-    }
+      ),
+    );
+
+    if (isAuthEnabled)
+      workspaceValidator.validateAuth(isAuthEnabled, workspace);
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts:200]: workspaceValidator.validateAuth(() => { isValid(authProvider), ${authProvider} not enabled)

File: packages/twenty-server/src/engine/core-modules/user/services/user.service.ts [hunk 13]
```diff
@@ -131,4 +134,29 @@ export class UserService extends TypeOrmQueryService<User> {
 
     return user;
   }
+
+  async saveDefaultWorkspace(userId: string, workspaceId: string) {
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/user/services/user.service.ts:138]: give a more explicit name

Comment (charlesBochet) [reply packages/twenty-server/src/engine/core-modules/user/services/user.service.ts:138]: saveDefautWorkspaceIfUserHasAccessOrThrow

Comment (charlesBochet) [reply packages/twenty-server/src/engine/core-modules/user/services/user.service.ts:138]: or extract hasUserAccessToWorkspace

File: packages/twenty-server/src/engine/core-modules/user/user.validate.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
+import { User } from 'src/engine/core-modules/user/user.entity';
+import { CustomException } from 'src/utils/custom-exception';
+
+const assertIsExist = (
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/user/user.validate.ts:4]: assertIsDefinedOrThrow

File: packages/twenty-server/src/engine/core-modules/user/user.validate.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+  user: User,
+  exceptionToThrow?: CustomException,
+): asserts user is User & { defaultWorkspaceId: string } => {
+  if (!user.defaultWorkspaceId) {
+    throw exceptionToThrow;
+  }
+};
+
+export const userValidator: {
+  assertIsExist: typeof assertIsExist;
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/user/user.validate.ts:27]: we would avoid typeof usage usually to lower complexity

File: packages/twenty-server/src/engine/core-modules/workspace-invitation/services/workspace-invitation.service.ts [hunk 16]
```diff
@@ -55,28 +147,41 @@ export class WorkspaceInvitationService {
...
+        workspaceId: workspace.id,
+        type: AppTokenType.InvitationToken,
+        deletedAt: IsNull(),
+      },
+      select: {
+        value: false,
+      },
+    });
+
+    return appTokens.map(castAppTokenToWorkspaceInvitationUtil);
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/workspace-invitation/services/workspace-invitation.service.ts:181]: move cast to an util

File: packages/twenty-server/src/engine/core-modules/workspace/workspace.resolver.ts [hunk 17]
```diff
@@ -144,4 +152,26 @@ export class WorkspaceResolver {
   hasValidEntrepriseKey(): boolean {
     return isDefined(this.environmentService.get('ENTERPRISE_KEY'));
   }
+
+  @Query(() => PublicWorkspaceDataOutput)
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/workspace/workspace.resolver.ts:156]: we could extract that to another resolver

File: packages/twenty-server/src/engine/core-modules/workspace/workspace.validate.ts [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+type IsAuthEnabled = <P extends WorkspaceAuthProvider>(
+  provider: P,
+  exceptionToThrow: CustomException,
+) => (
+  workspace: Workspace,
+  exceptionToThrowCustom?: CustomException,
+) => boolean;
+
+const isAuthEnabled: IsAuthEnabled = (provider, exceptionToThrow) => {
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/workspace/workspace.validate.ts:36]: see comments above

**Review by charlesBochet — APPROVED** [review]

File: packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts [hunk 19]
```diff
@@ -297,11 +297,16 @@ export class SignInUpService {
...
+      // let the creation of the first workspace
+      if (numberOfWorkspaces > 0) {
+        throw new AuthException(
+          'New workspace setup is disabled',
+          AuthExceptionCode.FORBIDDEN_EXCEPTION,
+        );
+      }
     }
 
     const workspaceToCreate = this.workspaceRepository.create({
```
[the remark this answers is not in the document]
Comment (zeckdude) [reply packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts]: What do you mean? Please expand.

[the remark this answers is not in the document]
Comment (zeckdude) [reply packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts]: @greptileai What do you mean? Please expand.

[the remark this answers is not in the document]
Comment (zeckdude) [reply packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts]: @greptile-apps What do you mean? Please expand.

**Pull request merged**